contracts/marketpools.js

```diff
@@ -0,0 +1,348 @@
+    || !api.assert(typeof (tokenSymbol) === 'string', 'invalid token')
+    || !api.assert(tokenOut && api.BigNumber(tokenOut).gt(0), 'insufficient tokenOut')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  let liquidityIn;
+  let liquidityOut;
+  let symbolIn;
+  let symbolOut;
+  if (tokenSymbol !== baseSymbol) {
+    liquidityIn = pool.baseQuantity;
+    liquidityOut = pool.quoteQuantity;
+    symbolIn = baseSymbol;
+    symbolOut = quoteSymbol;
+  } else {
+    liquidityIn = pool.quoteQuantity;
+    liquidityOut = pool.baseQuantity;
+    symbolIn = quoteSymbol;
+    symbolOut = baseSymbol;
+  }
+  if (!api.assert(pool, 'no existing pool for tokenPair')) return;
+
+  const tokenInAdjusted = api.BigNumber(getAmountIn(tokenOut, liquidityIn, liquidityOut));
+  if (!tokenInAdjusted.isFinite()) return;
+
+  const senderBase = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: symbolIn });
+  const senderBaseFunded = api.BigNumber(senderBase.balance).gte(tokenInAdjusted);
+  const tokenPairDelta = tokenSymbol === baseSymbol ? [api.BigNumber(tokenOut).times(-1), tokenInAdjusted] : [tokenInAdjusted, api.BigNumber(tokenOut).times(-1)];
+  if (!api.assert(senderBaseFunded, 'insufficient input balance')
+    || !validateSwap(pool, tokenPairDelta[0], tokenPairDelta[1])) return;
+
+  await api.executeSmartContract('tokens', 'transferToContract', { symbol: symbolIn, quantity: tokenInAdjusted.toFixed(pool.precision), to: 'marketpools' });
+  await api.transferTokens(api.sender, symbolOut, tokenOut, 'user');
+  updatePoolStats(pool, tokenPairDelta[0], tokenPairDelta[1]);
+  api.emit('swapTokensForExactTokens', { memo: `Swap ${symbolIn} for ${symbolOut}` });
+};
+
+actions.swapExactTokensForTokens = async (payload) => {
+  const {
+    tokenPair,
+    tokenSymbol,
+    tokenIn,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    || !api.assert(typeof (tokenSymbol) === 'string', 'invalid token')
+    || !api.assert(tokenIn && api.BigNumber(tokenIn).gt(0), 'insufficient tokenIn')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  let liquidityIn;
+  let liquidityOut;
+  let symbolIn;
+  let symbolOut;
+  if (tokenSymbol === baseSymbol) {
+    liquidityIn = pool.baseQuantity;
+    liquidityOut = pool.quoteQuantity;
+    symbolIn = baseSymbol;
+    symbolOut = quoteSymbol;
+  } else {
+    liquidityIn = pool.quoteQuantity;
+    liquidityOut = pool.baseQuantity;
+    symbolIn = quoteSymbol;
+    symbolOut = baseSymbol;
+  }
+
+  if (!api.assert(pool, 'no existing pool for tokenPair')) return;
```

Comment:
this assertion should be moved up to right after the findOne statement.

Reply:
Fixed